File: week#5/Vehicle.java
// Implement method overriding examples
// finding speed of a specific model using polymorphism
/* this program uses dynamic dispatch as its the necesscity of run-time polymorphism.
Dynamic Dispatch is a method in which the method that is to run in inheritance classes is decided in runtime.
 */
// public class
public class Vehicle {
    int speed;
    // parent method
    void showSpeed() {
        System.out.print("Speed of Vehicle depends on type & company.");
    }

    // main methid
    public static void main(String[] args) {
    // creating object of car in Vehicle class
        Vehicle car = new Car(100);
    // creating object of Honda in Car class
        Car honda = new Honda(70);
    // calling run overriden method
        car.showSpeed();
        honda.showSpeed();
    }

}
// child class
class Car extends Vehicle {

    Car(int speed){
        this.speed = speed;
    }

    // overriden method
    @Override
    void showSpeed() {
        System.out.println("Car max speed is around " + speed + "kmh.");
    }
}

class Honda extends Car {

    Honda(int speed) {
        super(speed);
    }

    // overriding the overridden method
    @Override
    void showSpeed() {
        System.out.println("Honda max speed is around " + speed + "kmh.");
    }
}

File: week#5/Vehicle.cpp
// Implement method overriding examples
// finding speed of a specific model using polymorphism
// Dynamic dispatch in C++ is done by vTables, This is achieved by creating a reference of object in a base class pointing to the child class. Enabling overriding via C++ by virtual methods (using virtual keyword), that allows it to execute method not based on reference type rather where the object is i.e., child class, this is done during runtime, This is called runtime polymorphism.

#include<iostream>
using std::string;
using std::cout;
using std::cin;

class Vehicle {
public:
    int speed;
    virtual void showSpeed() {
        cout<<"Vehice max speed depends on type, model & company.";
    }

};

class Car : public Vehicle {
public:
    Car(int speed) {
        this->speed = speed;
    }

     void showSpeed() override{
        cout<<"Car max speed is around "<<speed<<"kmh.\n";
    }

};

class Honda : public Car {
public:
    Honda(int speed) : Car(speed) {}
    void showSpeed() override {
        cout<<"Honda max speed is around "<<speed<<"kmh.";
    }    

};



int main() {
    Vehicle* car = new Car(200);
    car->showSpeed();
    Car* honda = new Honda(100);
    honda->showSpeed();

}